README.md
[![CI](https://github.com/bihealth/gene-ranking-shootout/actions/workflows/main.yml/badge.svg)](https://github.com/bihealth/gene-ranking-shootout/actions/workflows/main.yml)
[![Pypi Package Version](https://img.shields.io/pypi/v/gene-ranking-shootout.svg)](https://pypi.org/project/gene-ranking-shootout)
[![Pypi Python Version](https://img.shields.io/pypi/pyversions/gene-ranking-shootout.svg)](https://pypi.org/project/gene-ranking-shootout)

# Gene Ranking Shootout

A benchmark for methods that rank genes according to their relevance for a given phenotype (list of HPO terms).

## Methods

The following methods are currently included in the benchmark:

- AMELIE (via web service)
- CADA (via custom Docker/Podman image)
- Phen2Gene (via official Docker/Podman image)
- Phenix algorithm (as implemented in VarFish)
- Exomiser Algorithms:
  - Phenix
  - Phive
  - HiPhive (variants: human only, human-mouse only, all human-mouse-fish-ppi)

## Installation

Simply install with `pip` (probably inside a conda environment or virtualenv):

```bash
$ git clone https://github.com/bihealth/gene-ranking-shootout.git
$ cd gene-ranking-shootout
$ pip install -e .
$ gene-ranking-shootout --help
```

## Usage

List datasets in the benchmark:

```bash
$ gene-ranking-shootout dataset list
cada_clinvar_cases
cada_cases_test
cada_cases_validate
cada_all_cases
cada_collaborator_cases
cada_cases_train
```

Show the first entry in a dataset:

```bash
$ gene-ranking-shootout dataset head cada_cases_test --count 2
{"name": "Patient:SCV000281758", "disease_omim_id": "OMIM:617360", "disease_gene_id": "Entrez:8621", "hpo_terms": ["HP:0001508"], "candidate_gene_ids": null}
{"name": "Patient:SCV000864231", "disease_omim_id": "OMIM:132900", "disease_gene_id": "Entrez:4629", "hpo_terms": ["HP:0011499", "HP:0000021"], "candidate_gene_ids": null}
```

Simulate cases based on one or more datasets.
This will pick a number of cases from the datasets.
Further, it will pick another number of random genes based on the number of rare variants (freq below 0.1% in gnomAD genomes).
The results are written into a JSON file with the cases.
The simulation is randomized with a fixed seed that can be adjusted on the command line if necessary.

```bash
$ gene-ranking-shootout dataset simulate \
    /tmp/cases.json \
    $(gene-ranking-shootout dataset list) \
    --case-count 4714 \
    --candidate-genes-count 100
2023-05-05 11:16:35 | INFO   | Loading data
2023-05-05 11:16:35 | INFO   | ... 4714 cases overall (9428 duplicates)
2023-05-05 11:16:35 | INFO   | Simulating cases
2023-05-05 11:16:37 | INFO   | Wrote 4714 cases
```

You can then run the benchmark on the cases with the different methods:

```
$ gene-ranking-shootout benchmark
Usage: gene-ranking-shootout benchmark [OPTIONS] COMMAND [ARGS]...

  Group for benchmark sub commands.

Options:
  --help  Show this message and exit.

Commands:
  amelie          Benchmark the AMELIE web server.
  phen2gene       Benchmark the Phen2Gene container.
  summarize       Summarize the results.
  varfish-phenix  Benchmark the VarFish implementation of the Phenix...
$ gene-ranking-shootout benchmark amelie /tmp/cases.json /tmp/result-amelie.json
$ gene-ranking-shootout benchmark phen2gene /tmp/cases.json /tmp/result-phen2gene.json
$ gene-ranking-shootout benchmark varfish-phenix http://127.0.0.1:8081/hpo/sim/term-gene /tmp/cases.json /tmp/result-varfish-phenix.json
$ gene-ranking-shootout benchmark cada /tmp/cases.json /tmp/result-cada.json
$ gene-ranking-shootout benchmark exomiser http://localhost:8081/ phenix /tmp/cases.json /tmp/result-exomiser-phenix.json
$ gene-ranking-shootout benchmark exomiser http://localhost:8081/ phive /tmp/cases.json /tmp/result-exomiser-phive.json
$ gene-ranking-shootout benchmark exomiser http://localhost:8081/ hiphive /tmp/cases.json /tmp/result-exomiser-hiphive.json
$ gene-ranking-shootout benchmark exomiser http://localhost:8081/ hiphive-mouse /tmp/cases.json /tmp/result-exomiser-hiphive-mouse.json
$ gene-ranking-shootout benchmark exomiser http://localhost:8081/ hiphive-human /tmp/cases.json /tmp/result-exomiser-hiphive-human.json
```

You can also visualize the details of the benchmark results for each result file (below for 100 cases). This visualization displays the number of true disease genes (from case set definitions) at TOP10 and following positions in the ranked gene list of the respective method.

```bash
$ gene-ranking-shootout benchmark summarize /tmp/result-amelie.json
    1:   48  ################################
    2:   17  ###########
    3:    7  ####
    4:    3  ##
    5:    6  ####
    6:    3  ##
    7:    2  #
    8:    3  ##
    9:    1  .
   10:    0  

11-..:    9  ######
mssng:    0  
```

## Building CADA Podman Image

There is no public REST API or docker image for CADA (yet).
Here is how to build the needed CADA Podman image:

```bash
# cd docker/cada
# bash build.sh
...
# podman run -it --rm localhost/cada-for-shootout:latest --help
```

## Running Phenix in VarFish

Send an email to the author to get a copy of the necessary data.
Then, run the following in the background.

```bash
$ varfish-server-worker server pheno --path-hpo-dir path/to/varfish-server-worker-db/hpo
```

## Running Exomiser

The following are more rough notes than a full manual.
This uses the current Exomiser RES API version 13.2.0 (current at: 2023-05-05).
You will need approximately 75GB of storage for download and extraction and afterwards 49GB.
Probably one could get rid of a lot of the variant-specific data but I did not go into detail here.

```bash
$ wget https://github.com/exomiser/Exomiser/releases/download/13.2.0/exomiser-rest-prioritiser-13.2.0.jar
$ wget https://data.monarchinitiative.org/exomiser/latest/2302_phenotype.zip
$ wget https://data.monarchinitiative.org/exomiser/latest/2302_hg19.zip
$ unzip 2302_phenotype.zip
$ unzip 2302_hg19.zip
$ cat <<EOF > application.properties
exomiser.data-directory=$PWD
exomiser.hg19.data-version=1909
exomiser.phenotype.data-version=2302
exomiser.phenotype.random-walk-file-name=rw_string_10.mv
EOF
$ java -Xmx6G -Xms2G -Dserver.address=0.0.0.0 -Dserver.port=8081 -jar exomiser-rest-prioritiser-13.2.0.jar
```

## Datasets

The following datasets are included at the moment:

- `cada_cases_test.json` - converte from CADA's `cases_test.tsv`
- `cada_cases_train.json` - converte from CADA's `cases_train.tsv`
- `cada_cases_validate.json` - converte from CADA's `cases_validate.tsv`
- `cada_clinvar_cases.json` - converte from CADA's `clinvar_cases.tsv`
- `cada_collaborator_cases.json` - converte from CADA's `collaborator_cases.tsv`

You can conver TSV files with the following structure with `gene-ranking-shootout dataset convert-tsv`.

- Column 1: name for the case; must start with `Patient:` or is ignored.
- Column 2: disease_omim_id; as `OMIM:123456` or `unknown`
- Column 3: disease_gene_id; as `Entrez:123456`
- Column 4: hpo_terms; as semicolon-separated list, e.g., `HP:0001234;HP:0005678`

If a row has less than 4 columns, we assume that column 2 is missing.
All further columns are ignored.
The file should not have a header.
You can find some files in the CADA repository here:

- https://github.com/Chengyao-Peng/CADA/tree/main/src/CADA

The call to `gene-ranking-shootout dataset convert-tsv` should be as follows.

```bash
$ gene-ranking-shootout dataset convert-tsv input.tsv output.json
```

## Some Preliminary Results

The following was generated on 2023/05/05 with all 4714 cases.

```
$ for f in /tmp/result-*.json; do (set -x; gene-ranking-shootout benchmark summarize --bars-top-n 20 $f); echo; done
+ gene-ranking-shootout benchmark summarize --bars-top-n 20 result-cada.json
    1: 3462  ################################################
    2:  536  #######
    3:  205  ##
    4:  133  #
    5:   71  .
    6:   57  .
    7:   39  .
    8:   47  .
    9:   16  .
   10:   24  .
   11:   14  .
   12:   22  .
   13:   14  .
   14:   11  .
   15:    7  .
   16:    9  .
   17:    8  .
   18:    4  .
   19:    3  .
   20:    3  .

21-..:   29  .
mssng:    0  

+ gene-ranking-shootout benchmark summarize --bars-top-n 20 result-exomiser-hiphive-human.json
    1: 2593  ####################################
    2:  637  ########
    3:  375  #####
    4:  203  ##
    5:  132  #
    6:  104  #
    7:   99  #
    8:   80  #
    9:   61  .
   10:   46  .
   11:   44  .
   12:   30  .
   13:   30  .
   14:   30  .
   15:   15  .
   16:   22  .
   17:   12  .
   18:   11  .
   19:   13  .
   20:   10  .

21-..:  149  ##
mssng:    0  

+ gene-ranking-shootout benchmark summarize --bars-top-n 20 result-exomiser-hiphive.json
    1: 2418  #################################
    2:  686  #########
    3:  355  ####
    4:  226  ###
    5:  155  ##
    6:  122  #
    7:   86  #
    8:   95  #
    9:   54  .
   10:   54  .
   11:   52  .
   12:   28  .
   13:   29  .
   14:   27  .
   15:   16  .
   16:   11  .
   17:   23  .
   18:    9  .
   19:   10  .
   20:   14  .

21-..:  226  ###
mssng:    0  

+ gene-ranking-shootout benchmark summarize --bars-top-n 20 result-exomiser-hiphive-mouse.json
    1: 2418  #################################
    2:  685  #########
    3:  357  #####
    4:  227  ###
    5:  160  ##
    6:  121  #
    7:   90  #
    8:   96  #
    9:   55  .
   10:   56  .
   11:   58  .
   12:   37  .
   13:   31  .
   14:   26  .
   15:   22  .
   16:   18  .
   17:   26  .
   18:   19  .
   19:   14  .
   20:   15  .

21-..:  165  ##
mssng:    0  

+ gene-ranking-shootout benchmark summarize --bars-top-n 20 result-exomiser-phenix.json
    1: 2566  ####################################
    2:  619  ########
    3:  306  ####
    4:  208  ##
    5:  152  ##
    6:  119  #
    7:   90  #
    8:   80  #
    9:   73  #
   10:   64  .
   11:   49  .
   12:   41  .
   13:   40  .
   14:   26  .
   15:   29  .
   16:   27  .
   17:   15  .
   18:   17  .
   19:   10  .
   20:   16  .

21-..:  149  ##
mssng:    0  

+ gene-ranking-shootout benchmark summarize --bars-top-n 20 result-exomiser-phive.json
    1:  934  #############
    2:  298  ####
    3:  163  ##
    4:  101  #
    5:   50  .
    6:   44  .
    7:   30  .
    8:   33  .
    9:   16  .
   10:   16  .
   11:   10  .
   12:    9  .
   13:   17  .
   14:   13  .
   15:   12  .
   16:   16  .
   17:   14  .
   18:   23  .
   19:   28  .
   20:   33  .

21-..: 2836  #######################################
mssng:    0  

+ gene-ranking-shootout benchmark summarize --bars-top-n 20 result-phen2gene.json
    1: 2426  ##################################
    2:  470  ######
    3:  209  ##
    4:  125  #
    5:  101  #
    6:   67  .
    7:   51  .
    8:   62  .
    9:   53  .
   10:   41  .
   11:   33  .
   12:   37  .
   13:   42  .
   14:   33  .
   15:   34  .
   16:   28  .
   17:   18  .
   18:   29  .
   19:   17  .
   20:   19  .

21-..:  763  ##########
mssng:    0  

+ gene-ranking-shootout benchmark summarize --bars-top-n 20 result-varfish-phenix.json
    1: 1709  #######################
    2:  616  ########
    3:  357  ####
    4:  277  ###
    5:  184  ##
    6:  152  ##
    7:  131  #
    8:  118  #
    9:  105  #
   10:   78  #
   11:   71  .
   12:   57  .
   13:   67  .
   14:   64  .
   15:   67  .
   16:   71  .
   17:   56  .
   18:   48  .
   19:   34  .
   20:   48  .

21-..:  403  #####
mssng:    0  
```


CHANGELOG.md
# Changelog

### [0.1.3](https://www.github.com/bihealth/gene-ranking-shootout/compare/v0.1.2...v0.1.3) (2023-05-30)


### Bug Fixes

* filter CADA results to candidate gene IDs ([#22](https://www.github.com/bihealth/gene-ranking-shootout/issues/22)) ([#23](https://www.github.com/bihealth/gene-ranking-shootout/issues/23)) ([eb65c72](https://www.github.com/bihealth/gene-ranking-shootout/commit/eb65c72f22ee8fc0632e6e5416edfacf58736482))


### Documentation

* adding new methods to README ([#18](https://www.github.com/bihealth/gene-ranking-shootout/issues/18)) ([8e5c82d](https://www.github.com/bihealth/gene-ranking-shootout/commit/8e5c82dfae243ef7f4ecf2aea4cf2e1b0ad9964f))
* Update README.md with details on 'benchmark summarize' output ([#21](https://www.github.com/bihealth/gene-ranking-shootout/issues/21)) ([3904c92](https://www.github.com/bihealth/gene-ranking-shootout/commit/3904c92349f4f1a84af3d28c6cdb0a35cbde6b25))

### [0.1.2](https://www.github.com/bihealth/gene-ranking-shootout/compare/v0.1.1...v0.1.2) (2023-05-05)


### Documentation

* fixing README in setup.cfg, adding PyPi badges ([#15](https://www.github.com/bihealth/gene-ranking-shootout/issues/15)) ([16d4f32](https://www.github.com/bihealth/gene-ranking-shootout/commit/16d4f32b94195a85df06ad6bddb85a6381c33bfe))

### [0.1.1](https://www.github.com/bihealth/gene-ranking-shootout/compare/v0.1.0...v0.1.1) (2023-05-05)


### Build System

* properly use git version with setuptools_scm ([#13](https://www.github.com/bihealth/gene-ranking-shootout/issues/13)) ([2434052](https://www.github.com/bihealth/gene-ranking-shootout/commit/243405280d799f0e62157ebc7baa62677cb4ecdc))

## 0.1.0 (2023-05-05)


### Features

* add AMELIE to benchmark ([#6](https://www.github.com/bihealth/gene-ranking-shootout/issues/6)) ([eb7fd5c](https://www.github.com/bihealth/gene-ranking-shootout/commit/eb7fd5ceacb9138e4bd4e5a1eb6988e0cd9f21c5))
* add support for exomiser ([#11](https://www.github.com/bihealth/gene-ranking-shootout/issues/11)) ([4ca3587](https://www.github.com/bihealth/gene-ranking-shootout/commit/4ca3587994b748a79c96acaea5f0c8b3ab4b22db))
* adding CADA to benchmark ([#7](https://www.github.com/bihealth/gene-ranking-shootout/issues/7)) ([237afb5](https://www.github.com/bihealth/gene-ranking-shootout/commit/237afb5283d2474a2a744ef536603725e395f458))
* adding more data ([#9](https://www.github.com/bihealth/gene-ranking-shootout/issues/9)) ([d3906ac](https://www.github.com/bihealth/gene-ranking-shootout/commit/d3906aca4008d553daf623c99ef2c5ce957a860b))
* adding phen2gene benchmark ([#5](https://www.github.com/bihealth/gene-ranking-shootout/issues/5)) ([e701271](https://www.github.com/bihealth/gene-ranking-shootout/commit/e701271cb94b70db24645e50e3193b43faa5b053))
* adding some data commands ([#1](https://www.github.com/bihealth/gene-ranking-shootout/issues/1)) ([ab702d5](https://www.github.com/bihealth/gene-ranking-shootout/commit/ab702d57a521d6194263511599b4fbd48b016fde))
* benchmark for phenix algo in varfish ([#3](https://www.github.com/bihealth/gene-ranking-shootout/issues/3)) ([0dc2c77](https://www.github.com/bihealth/gene-ranking-shootout/commit/0dc2c77b6c549bfcde9743e09e94627e9398ed3b))


### Bug Fixes

* parameters to hiphive ([#12](https://www.github.com/bihealth/gene-ranking-shootout/issues/12)) ([a4ebd88](https://www.github.com/bihealth/gene-ranking-shootout/commit/a4ebd885b2eb43cd5d5da326b9ce25b006daca0b))


### Documentation

* adding more results to README ([#10](https://www.github.com/bihealth/gene-ranking-shootout/issues/10)) ([afdb1b4](https://www.github.com/bihealth/gene-ranking-shootout/commit/afdb1b4e594686035c93c806947c4d6b0dce30ca))
* extending the README ([#8](https://www.github.com/bihealth/gene-ranking-shootout/issues/8)) ([9b7a6ad](https://www.github.com/bihealth/gene-ranking-shootout/commit/9b7a6ad724db99821fa8badbf30f5cd846fbb32b))

## 0.1.0 (2022-12-02)


gene_ranking_shootout/models.py
"""Data models"""

import csv
import json
import pathlib
import typing

import attrs


@attrs.frozen()
class Case:
    """A case for benchmarking."""

    #: The case's name.
    name: str
    #: The OMIM code of the disease, e.g., ``"OMIM:251110"``.
    disease_omim_id: str
    #: The disease gene ID, e.g., ```"Entrez:1301"`.
    disease_gene_id: str
    #: The HPO terms for describing the patient, e.g. ``["HP:0001263", "HP:0000256",
    #: "HP:0008414"].
    hpo_terms: typing.List[str]
    #: Candidate gene IDs (e.g., ``["Entrez:1301", "Entrez:1302"]``).
    candidate_gene_ids: typing.Optional[typing.List[str]] = None


def load_cases_json(path):
    """Load ``Case`` objects from JSON file."""
    with open(path, "rt") as f:
        return [Case(**case) for case in json.load(f)]


@attrs.frozen()
class GnomadCounts:
    """Per-gene count data for gnomAD"""

    #: HGNC gene symbol.
    gene_symbol: str
    #: HGNC gene ID.
    hgnc_id: str
    #: Entrez gene ID.
    entrez_id: str
    #: Rare variant count in gnomAD.
    count: int


def load_gnomad_counts_tsv(path):
    """Load ``GnomadCounts`` objects from TSV file (with header)."""
    with open(path, "rt") as f:
        reader = csv.DictReader(f, delimiter="\t")
        return [GnomadCounts(**row) for row in reader]


def load_gnomad_counts():
    """Helper to load the gnomad counts from the data directory."""
    return load_gnomad_counts_tsv(pathlib.Path(__file__).parent / "data" / "gnomad_counts.tsv")


@attrs.frozen()
class Result:
    """Stores prioritization results."""

    #: The case that was run.
    case: Case
    #: The rank that the disease gene was assigned.
    rank: int
    #: The resulting ranked genes as Entrez IDs.
    result_entrez_ids: typing.List[str]


gene_ranking_shootout/cli/__init__.py
"""CLI interface"""

import csv
import json
import pathlib
import sys
import typing

import attrs
import cattrs
import click
from loguru import logger
import numpy as np

from gene_ranking_shootout import models, runner


@click.group()
def main():
    """Main entry point for the CLI interface"""
    logger.remove()
    fmt = (
        "<green>{time:YYYY-MM-DD HH:mm:ss}</green> | "
        "<level>{level: <6}</level> | "
        "<level>{message}</level>"
    )
    logger.add(sys.stderr, format=fmt)


@main.group()
def benchmark():
    """Group for benchmark sub commands."""


@benchmark.command()
@click.option("--bars-top-n", default=10)
@click.option("--total-width", default=80)
@click.argument("results_json")
def summarize(results_json, bars_top_n, total_width):
    """Summarize the results."""
    with open(results_json, "rt") as inf:
        results = cattrs.structure(json.load(inf), typing.List[models.Result])
    runner.BarPrinter(bars_top_n=bars_top_n, total_width=total_width).print(results)


@benchmark.command()
@click.option("--bars-top-n", default=10)
@click.option("--total-width", default=80)
@click.argument("base_url")
@click.argument("simulated_json")
@click.argument("results_json")
def varfish_phenix(base_url, simulated_json, results_json, bars_top_n, total_width):
    """Benchmark the VarFish implementation of the Phenix algorithm."""
    runner.PhenixVarFishRunner(base_url, bars_top_n=bars_top_n, total_width=total_width).run(
        simulated_json, results_json
    )


@benchmark.command()
@click.option("--bars-top-n", default=10)
@click.option("--total-width", default=80)
@click.argument("simulated_json")
@click.argument("results_json")
def phen2gene(simulated_json, results_json, bars_top_n, total_width):
    """Benchmark the Phen2Gene container."""
    runner.Phen2GeneRunner(bars_top_n=bars_top_n, total_width=total_width).run(
        simulated_json, results_json
    )


@benchmark.command()
@click.option("--bars-top-n", default=10)
@click.option("--total-width", default=80)
@click.argument("simulated_json")
@click.argument("results_json")
def amelie(simulated_json, results_json, bars_top_n, total_width):
    """Benchmark the AMELIE web server."""
    runner.AmelieRunner(bars_top_n=bars_top_n, total_width=total_width).run(
        simulated_json, results_json
    )


@benchmark.command()
@click.option("--bars-top-n", default=10)
@click.option("--total-width", default=80)
@click.option("--threads", default=0)
@click.argument("simulated_json")
@click.argument("results_json")
def cada(simulated_json, results_json, bars_top_n, total_width, threads):
    """Benchmark the CADA container."""
    runner.CadaRunner(bars_top_n=bars_top_n, total_width=total_width, threads=threads).run(
        simulated_json, results_json
    )


@benchmark.command()
@click.option("--bars-top-n", default=10)
@click.option("--total-width", default=80)
@click.argument("base_url")
@click.argument("algorithm")
@click.argument("simulated_json")
@click.argument("results_json")
def exomiser(base_url, algorithm, simulated_json, results_json, bars_top_n, total_width):
    """Benchmark the Exomiser REST Prioritizer."""
    runner.ExomiserRunner(base_url, algorithm, bars_top_n=bars_top_n, total_width=total_width).run(
        simulated_json, results_json
    )


@main.group()
def dataset():
    """Group for dataset sub commands."""


@dataset.command("list")
def list_():
    """Listing of available data sets."""
    # List all files ending in .json in the data directory
    data_dir = pathlib.Path(__file__).parent.parent / "data"
    for path in data_dir.glob("*.json"):
        print(path.stem)


def load_dataset(dataset):
    data_dir = pathlib.Path(__file__).parent.parent / "data"
    path = data_dir / f"{dataset}.json"
    return models.load_cases_json(path)


@dataset.command()
@click.argument("dataset")
@click.option("--count", default=10)
def head(dataset, count):
    """Display first entry in the given dataset."""
    cases = load_dataset(dataset)
    for case in cases[:count]:
        print(json.dumps(cattrs.unstructure(case)))


@dataset.command()
@click.argument("out_json")
@click.argument("datasets", nargs=-1)
@click.option("--seed", default=42)
@click.option("--case-count", default=10)
@click.option("--candidate-genes-count", default=19)
def simulate(out_json, datasets, case_count, candidate_genes_count, seed):
    """Simulate cases based on the dataset file."""
    # Load dataset and all genes from ``gnomad_counts.tsv``.
    logger.info("Loading data")
    cases = []
    skipped = 0
    seen_case_names = set()
    for dataset in datasets:
        for case in load_dataset(dataset):
            if case.name in seen_case_names:
                skipped += 1
                continue
            cases.append(case)
            seen_case_names.add(case.name)
    logger.info("... {} cases overall ({} duplicates)", len(cases), skipped)
    gnomad_counts = models.load_gnomad_counts()
    gnomad_counts_idx = np.array(range(len(gnomad_counts)))
    gnomad_counts_values = np.array([gene.count for gene in gnomad_counts]).astype(np.float64)
    gnomad_counts_sum = np.sum(gnomad_counts_values)
    gnomad_counts_ps = np.divide(gnomad_counts_values, gnomad_counts_sum)
    logger.info("Simulating cases")
    # Create random number generator.
    rng = np.random.default_rng(seed)
    # Pick cases
    cases_idxs = np.array(range(len(cases)))
    cases_idx = rng.choice(cases_idxs, size=case_count, replace=False)
    selected_cases = [cases[idx] for idx in cases_idx]
    # For each case, pick random candidate genes.
    simulated = []
    for case in selected_cases:
        candidate_gene_idxs = rng.choice(
            gnomad_counts_idx, size=candidate_genes_count + 1, replace=False, p=gnomad_counts_ps
        ).tolist()
        candidate_genes = [gnomad_counts[idx] for idx in candidate_gene_idxs]
        if case.disease_gene_id in candidate_genes:
            candidate_genes.remove(case.disease_gene_id)
        else:
            candidate_genes.pop()
        rng.shuffle(candidate_genes)
        simulated.append(
            attrs.evolve(case, candidate_gene_ids=[gene.entrez_id for gene in candidate_genes])
        )
    # Write result to JSON file.
    with open(out_json, "wt") as f:
        json.dump(cattrs.unstructure(simulated), f, indent=2)
    logger.info("Wrote {} cases", len(simulated))


@dataset.command()
@click.argument("tsv_in")
@click.argument("json_out")
def convert_tsv(tsv_in, json_out):
    """Convert from TSV to JSON format."""
    logger.info("Converting from {} to {}", tsv_in, json_out)
    with open(tsv_in, "rt") as inputf:
        with open(json_out, "wt") as outputf:
            cases = []
            header = ("name", "disease_omim_id", "disease_gene_id", "hpo_terms")
            reader = csv.reader(inputf, delimiter="\t")
            for row in reader:
                if not row[0].startswith("Patient"):
                    logger.info("Skipping row {}; does not start with 'Patient:'", row)
                    continue
                elif len(row) < 3:
                    logger.info("Skipping row {}; not enough columns", row)
                    continue
                if len(row) == 3:
                    data = dict(zip(header[:1] + header[2:], row))
                else:
                    data = dict(zip(header, row))
                cases.append(
                    models.Case(
                        name=str(data["name"]),
                        disease_omim_id=str(data.get("disease_omim_id", "unknown")),
                        disease_gene_id=str(data["disease_gene_id"]),
                        hpo_terms=[str(term) for term in data["hpo_terms"].split(",")],
                    )
                )
            logger.info("Writing {} cases", len(cases))
            json.dump(cattrs.unstructure(cases), outputf, indent=2)


if __name__ == "__main__":
    main()
